lucene/core/src/java/org/apache/lucene/search/SliceExecutor.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import java.util.Collection;
+import java.util.concurrent.Executor;
+import java.util.concurrent.RejectedExecutionException;
+
+/**
+ * Executor which is responsible
+ * for execution of slices based on the current status
+ * of the system and current system load
+ */
+class SliceExecutor {
```

Comment:
What there a reason for not making `SliceExecutor` public class ?
I'd want to inject my own `SliceExecutor` without having to fork Lucene

Reply:
Not really. I will fix this and add to current release

Reply:
Great thanks a lot !

While I'm at it: 
the `reduce` operation also runs on the main thread, it would be lovely to make this configurable.

Glad to put up a PR for it if I find time. I'd hesitate between adding some more methods on `SliceExecutor`, or plumbing the `reduce` executor totally independently. 
The name `SliceExecutor` feels awkward if we also add logic for `reduce` there. Since `SliceExecutor` wasn't public, maybe we can still afford a rename ?